lib/spack/spack/directives.py

```diff
@@ -465,23 +465,54 @@ def variant(
         multi (bool): if False only one value per spec is allowed for
             this variant
         validator (callable): optional group validator to enforce additional
-            logic. It receives a tuple of values and should raise an instance
-            of SpackError if the group doesn't meet the additional constraints
+            logic. It receives the package name, the variant name and a tuple
+            of values and should raise an instance of SpackError if the group
+            doesn't meet the additional constraints
+
+    Raises:
+        DirectiveError: if arguments passed to the directive are invalid
     """
     if name in reserved_names:
-        raise ValueError("The variant name '%s' is reserved by Spack" % name)
+        raise DirectiveError(
+            "The variant name '%s' is reserved by Spack" % name
+        )
 
+    # Ensure we have a sequence of allowed variant values, or a
+    # predicate for it.
     if values is None:
-        if default in (True, False) or (type(default) is str and
-                                        default.upper() in ('TRUE', 'FALSE')):
+        if str(default).upper() in ('TRUE', 'FALSE'):
             values = (True, False)
         else:
             values = lambda x: True
 
-    if default is None:
-        default = False if values == (True, False) else ''
+    # The object defining variant values might supply its own defaults for
+    # all the other arguments. Ensure we have no conflicting definitions
+    # in place.
+    for argument in ('default', 'multi', 'validator'):
+        if hasattr(values, argument) and locals()[argument] is not None:
+            def _raise_argument_error(pkg):
+                msg = "multiple conflicting values for argument '{0}' " \
+                      "in variant '{1}' of package '{2}'"
+                raise DirectiveError(msg.format(argument, name, pkg.name))
+            return _raise_argument_error
+
+    # Allow for the object defining the allowed values to supply its own
+    # default value and group validator, say if it supports multiple values.
+    default = getattr(values, 'default', default)
+    validator = getattr(values, 'validator', validator)
+    multi = getattr(values, 'multi', bool(multi))
+
+    # Here we sanitize against a default value being either None
+    # or the empty string, as the former indicates that a default
+    # was not set while the latter will make the variant unparsable
+    # from the command line
+    if default is None or default == '':
+        def _raise_default_not_set(pkg):
+            msg = "the default value in variant '{0}' from package" \
+                  " '{1}' needs to be set explicitly"
```

Comment:
Where can I edit the docs to address these points:

> * How they code the possibility of choosing no value
> * How they specify this on the command line (e.g. when the default is not "none")

?

Reply:
@alalazo I missed this question:

My first thought was that there would be a section on variants in the package writing guide (https://spack.readthedocs.io/en/latest/packaging_guide.html#packaging-guide) and that would be the place to put it, but apparently there isn't any such section.

IMO this PR would ideally add a variants section to the packaging guide, and it would mention these conventions. If preferred, I can add a commit to this PR which adds that documentation.

Reply:
I should have time to add docs + push forward this PR over the week-end. Will ping you when done.